src/test/SampleCustomLoader/sampleCustLoader.cpp

```diff
@@ -0,0 +1,320 @@
+        cout << "custSampleLoader: Invalid custom loader options" << endl;
+        return CUSTLOADER_ERROR;
+    }
+
+    // load models
+    ret = load_files(binFile, xmlFile, xmlBuffer, binBuffer, xmlLen, binLen);
+    if (ret != CUSTLOADER_OK || xmlBuffer == NULL || xmlBuffer == NULL) {
+        cout << "custSampleLoader: Could not read model files" << endl;
+        return CUSTLOADER_ERROR;
+    }
+
+    /* Start the watcher thread after first moel load */
+    if (watcherStarted == false) {
+        int interval = 30;
+        startWatcher(interval);
+        this_thread::sleep_for(chrono::seconds(1));
+    }
+
+    models_loaded.push_back(make_pair(string(modelName), version));
+    return CUSTLOADER_OK;
+}
+
+int custSampleLoader::unloadModel(const char* modelName, int version) {
+    cout << "custSampleLoader: Custom unloadModel" << endl;
+
+    model_id_t toFind = make_pair(string(modelName), version);
+
+    auto it = models_loaded.begin();
+    for (; it != models_loaded.end(); it++) {
+        if (*it == toFind)
+            break;
+    }
+
+    if (it == models_loaded.end()) {
+        cout << modelName << " is not loaded" << endl;
+    } else {
+        models_loaded.erase(it);
+    }
+    return CUSTLOADER_OK;
+}
+
+void custSampleLoader::loaderDeInit() {
+    cout << "custSampleLoader: Custom loaderDeInit" << endl;
+    watcherJoin();
+    return;
+}
+
+bool custSampleLoader::getModelBlacklistStatus(const char* modelName, int version) {
+    cout << "custSampleLoader: Custom getModelBlacklistStatus" << endl;
+
+    if (models_blacklist.size() == 0)
+        return false;
+
+    model_id_t toFind = make_pair(string(modelName), version);
+
+    auto it = models_blacklist.begin();
+    for (; it != models_blacklist.end(); it++) {
+        if (*it == toFind)
+            break;
+    }
+
+    if (it == models_blacklist.end()) {
+        return false;
+    }
+
+    /* model name and version in blacklist.. return true */
+    return true;
+}
+
+void custSampleLoader::threadFunction(future<void> futureObj) {
+    cout << "custSampleLoader: Thread Start" << endl;
+    int count = 1;
+    while (futureObj.wait_for(chrono::milliseconds(1)) == future_status::timeout) {
+        cout << "custSampleLoader: Doing Some Work " << count++ << endl;
+        this_thread::sleep_for(std::chrono::seconds(watchIntervalSec));
+        // After 1 cycles forcing the thread to add first model as blacklisted
+        if (count == 2) {
+            if (models_loaded.size() > 0) {
+                models_blacklist.push_back(models_loaded[0]);
+                cout << "custSampleLoader: Blacklisting the model " << endl;
+            }
+        }
+
+        // After 2 cycles clear the blacklist and break
+        if (count == 3) {
+            cout << "custSampleLoader: Clearing the blacklist " << endl;
+            models_blacklist.clear();
+        }
+    }
+    cout << "custSampleLoader: Thread END" << endl;
+    watcherJoin();

```

Comment:
Possible race condition when this code executes before 
watcherStarted = true;
was executed in the main thread.

Reply:
This part is now fixed